ci_connection/preserve_run_state.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2024 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     https://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities for saving environment/execution state for use in SSH sessions.
+
+The setup/environment of a workflow can be saved, and later reproduced within
+an SSH session to the runner that is running the workflow in question.
+
+This is generally meant for debugging errors in CI.
+
+Can be used both as CLI, and library.
+"""
+
+import argparse
+import json
+import logging
+import os
+import re
+from typing import NotRequired, Sequence, TypedDict
+
+import utils
+
+
+utils.setup_logging()
+
+VARS_BLACKLIST = ("GITHUB_TOKEN",)
```

Comment:
I didn't mark it everywhere but the "blacklist" should be changed across the entire PR to be deny